Changes to AgentSetDF and AgentsDF before time.py -> CopyMixin

adamamer20 opened the pull request:
- Created a CopyMixin for copy and deepcopy methods to let both AgentSet and Scheduler inherit fast copy.
- Added possibility of using 'do' method with a mask in AgentSetDF, such that the Scheduler can use it to run the method only on a subset of agents (active agents in the scheduler).
- Added tests for AgentsDF
- Other minor bug fixes, including type hints and docstrings.

rht commented:
It looks like pre-commit.ci hasn't been enabled: https://pre-commit.ci/.

adamamer20 commented:
You're right, now it should be enabled.
So now before committing i should run locally `pre-commit run --all-files` right?

rht commented:
> So now before committing i should run locally pre-commit run --all-files right?

For now, yes, manually. Because it will only trigger starting in the next PR. You could also add a local Git commit hook so that it is automatic.

rht commented on mesa_frames/types.py:
```diff
@@ -1,20 +1,22 @@
+from collections.abc import Collection, Hashable
+from typing import Literal
+
+import pandas as pd
```
The problem with these imports is that user will have to install both pandas and Polars.

adamamer20 replied:
You're right, but I don't have a clear idea on how to type hint `DataFrame`, `Series `ecc. properly otherwise..

rht replied:
I assume putting the imports within the `if TYPE_CHECKING:` doesn't work on this file? If it doesn't work, then it's fine, package storage is not that big of an issue nowadays anyway.

adamamer20 replied:
Unfortunately it doesn't because type aliases are defined as variables importable from other modules. Maybe there's a better way to solve this but I need to think about it

adamamer20 commented:
@rht 
Ubuntu 3.10 is failing because typing.Self was introduced in Python 3.11. I have seen there is the possbility of using [`typing-extensions`](https://pypi.org/project/typing-extensions/) to allow typing hints for previous Python versions. Do you think it's worth it?
I am not sure why macos 3.12 is failing, the test passes in other builds.

adamamer20 commented:
@rht 
The ubuntu 3.10 build is failing because `typing.Self` was introduced in 3.11. I have seen that [this](https://pypi.org/project/typing-extensions/) can port typing features to previous Python versions. Do you think it's worth it?
Macos 3.12 is failing on a single test which seems to work on other platforms so i don't really understand the issue.
Also, obtaining a codecov token isn't free, right?

rht commented:
I tried adding `from __future__ import annotations` in agents.py. Didn't work. So I have removed my commit so that you are the sole author of this PR (otherwise it would get interpreted as 50% co-authorship without careful inspection).

Adding typing-extensions sounds good to me. Seems like a lightweight module.

adamamer20 commented:
@rht ready to merge

rht commented on tests/test_agents.py:
```diff
@@ -0,0 +1,976 @@
+
+        # Test with a single AgentSetPandas
+        result = agents.add(agentset_pandas, inplace=False)
+        assert result._agentsets[0] is agentset_pandas
```
You should only test via the public API's, because private structure may change all the time.

adamamer20 replied:
You're right, I thought that the public API would be more unstable, but as we continue development, it is the other way around. I will open an issue to change tests from private to public APIs in the future.

rht commented:
> Added possibility of using 'do' method with a mask in AgentSetDF, such that the Scheduler can use it to run the method only on a subset of agents (active agents in the scheduler).

Sorry for the late comment on this, but the scheduler (random activation, staged activation, etc, except for the discrete event scheduler) will be rendered obsolete, replaced with AgentSet's select-then-do.

But given that this PR has lots of other improvements, e.g. fixing the CI, I'm going to merge it, so that we can move faster. I need a working tests/CI in order to benchmark the runtime type checking libraries.

adamamer20 commented:
> Sorry for the late comment on this, but the scheduler (random activation, staged activation, etc, except for the discrete event scheduler) will be rendered obsolete, replaced with AgentSet's select-then-do.
> 
> But given that this PR has lots of other improvements, e.g. fixing the CI, I'm going to merge it, so that we can move faster. I need a working tests/CI in order to benchmark the runtime type checking libraries.

Ah, no problem! I thought that the scheduler would be kept and the AgentSet's select-then-do was only incentivized. Anyway, all the changes in the PR are still useful because masked do is something that is needed anyway. I have implemented the interface for schedulers (apart from the discrete event scheduler, which I was leaving to its own PR) in the `9-refactoring-mesatime-schedulers `branch. I still need to implement the tests, though. At this point, if you plan on deprecating soon, I will leave it as is and start working either on `mesa.space` or the discrete event scheduler. What do you think?

rht commented:
I think I'm more looking forward to the `mesa.space` first, as it will unlock us a more nontrivial example.
